crates/webidl/src/generator.rs

```diff
@@ -0,0 +1,818 @@
+        let mut required_args = vec![];
+        let mut required_calls = vec![];
+
+        for field in fields.iter() {
+            if field.required {
+                let name = &field.name;
+                let ty = &field.ty;
+                required_args.push(quote!( #name: #ty ));
+                required_calls.push(quote!( ret.#name(#name); ));
+                add_features(&mut required_features, &field.ty, &name.to_string());
+            }
+        }
+
+        let cfg_features = get_cfg_features(&required_features);
+
+        required_features.insert(name.to_string());
+
+        let doc_features = required_doc_string(&required_features);
+
+        let doc_comment = comment(format!("The `{}` dictionary.", name), &doc_features);
+        let ctor_doc_comment = comment(format!("Construct a new `{}`.", name), &doc_features);
+
+        let fields = fields.into_iter()
+            .map(|field| field.generate_rust(name.to_string()))
+            .collect::<Vec<_>>();
+
+        quote! {
+            use super::*;
+            use wasm_bindgen::prelude::*;
+
+            #unstable_attr
+            #[wasm_bindgen]
+            extern "C" {
+                #[wasm_bindgen(extends = ::js_sys::Object, js_name = #js_name)]
+                #[doc = #doc_comment]
+                pub type #name;
+            }
+
+            #unstable_attr
+            impl #name {
+                #cfg_features
+                #[doc = #ctor_doc_comment]
+                pub fn new(#(#required_args),*) -> Self {
+                    #[allow(unused_mut)]
+                    let mut ret: Self = ::wasm_bindgen::JsCast::unchecked_into(::js_sys::Object::new());
+                    #(#required_calls)*
+                    ret
+                }
+
+                #(#fields)*
+            }
+        }
+    }
+}
+
+
+pub struct Function {
+    pub name: Ident,
+    pub js_name: String,
+    pub arguments: Vec<(Ident, Type)>,
+    pub ret_ty: Option<Type>,
+    pub catch: bool,
+    pub variadic: bool,
+    pub unstable: bool,
+}
+
+impl Function {
+    fn generate(&self, parent_name: &Ident, parent_js_name: String) -> TokenStream {
+        let Function {
+            name,
+            js_name,
+            arguments,
+            ret_ty,
+            catch,
+            variadic,
+            unstable,
+        } = self;
+
+        let unstable_attr = maybe_unstable_attr(*unstable);
+
+        let js_namespace = raw_ident(&parent_js_name);
+
+        let doc_comment = format!(
+            "The `{}.{}()` function.\n\n{}",
+            parent_js_name,
+            js_name,
+            mdn_doc(&parent_js_name, Some(&js_name))
+        );
+
+        let mut features = BTreeSet::new();
+
+        for (_, ty) in arguments.iter() {
+            add_features(&mut features, ty, &parent_name.to_string());
+        }
+
+        if let Some(ty) = ret_ty {
+            add_features(&mut features, ty, &parent_name.to_string());
```

Comment:
This seems like a bit of a wonky API where `parent_name` is explicitly left out but then it's inserted just below. Could the `add_features` API avoid taking a parameter-to-avoid and then anyone who wants to remove it explicitly removes it afterwards?

Reply:
It is a bit wonky, but it's that way for a good reason: we want to exclude the `parent_name` from the `cfg`, but include it in the `doc`.

So it first adds the non-`parent_name` features, then it generates the `cfg` attribute, then it adds the `parent_name` feature, then it generates the `doc`.

I'm open to a better suggestion, but your suggestion would look like this:

```rust
let mut features = BTreeSet::new();

for (_, ty) in arguments.iter() {
    add_features(&mut features, ty);
}

if let Some(ty) = ret_ty {
    add_features(&mut features, ty);
}

features.remove(parent_name.to_string());

let cfg_features = get_cfg_features(&features);

features.insert(parent_name.to_string());

let doc_comment = comment(doc_comment, &required_doc_string(&features));
```

I'm not sure if that's an improvement.

Reply:
So, I realized that we actually need to remove multiple features, since the parent `interface` can have superclasses. So I changed it to remove the features and then re-add `parent_name`.